File: services/scene_detector.py
#!/usr/bin/env python3
"""
Scene detection script using pyscenedetect's ContentDetector.
Called as a subprocess from the Node.js application.
"""

import sys
import json
import argparse

try:
    from scenedetect import open_video, SceneManager
    from scenedetect.detectors import ContentDetector
    from scenedetect.backends import VideoStreamCv2
except ImportError:
    print(json.dumps({"error": "pyscenedetect not installed. Run: pip install scenedetect"}), file=sys.stderr)
    sys.exit(1)


def detect_scenes(video_path, threshold=30.0, min_scene_length_sec=0.5):
    """
    Detect scenes in a video using ContentDetector.
    
    Args:
        video_path: Path to the video file
        threshold: Threshold for content detection (0-255, default 30)
        min_scene_length_sec: Minimum scene length in seconds
    
    Returns:
        List of scene dictionaries with start/end frame and time info
    """
    try:
        # Open video
        video = open_video(video_path)
        
        # Get video properties
        fps = video.frame_rate
        total_frames = video.duration.get_frames() if hasattr(video.duration, 'get_frames') else int(video.duration * fps)
        duration = video.duration.get_seconds() if hasattr(video.duration, 'get_seconds') else video.duration
        
        # Convert min scene length from seconds to frames
        min_scene_length_frames = int(min_scene_length_sec * fps)
        
        # Ensure at least 1 frame
        if min_scene_length_frames < 1:
            min_scene_length_frames = 1
        
        # Create scene manager with ContentDetector
        scene_manager = SceneManager()
        scene_manager.add_detector(ContentDetector(
            threshold=threshold,
            min_scene_len=min_scene_length_frames
        ))
        
        # Detect scenes
        scene_manager.detect_scenes(video)
        
        # Get scene list
        scene_list = scene_manager.get_scene_list()
        
        # Convert to our format
        scenes = []
        for i, (start, end) in enumerate(scene_list, 1):
            start_frame = start.get_frames() if hasattr(start, 'get_frames') else int(start.get_seconds() * fps)
            end_frame = end.get_frames() if hasattr(end, 'get_frames') else int(end.get_seconds() * fps)
            start_time = start.get_seconds() if hasattr(start, 'get_seconds') else start
            end_time = end.get_seconds() if hasattr(end, 'get_seconds') else end
            
            scenes.append({
                "scene_number": i,
                "start_frame": start_frame + 1,  # 1-indexed
                "end_frame": end_frame,
                "start_time": round(start_time, 3),
                "end_time": round(end_time, 3),
                "duration": round(end_time - start_time, 3)
            })
        
        # If no scenes detected, treat whole video as one scene
        if not scenes:
            scenes.append({
                "scene_number": 1,
                "start_frame": 1,
                "end_frame": total_frames,
                "start_time": 0.0,
                "end_time": round(duration, 3),
                "duration": round(duration, 3)
            })
        
        return {
            "success": True,
            "scenes": scenes,
            "fps": fps,
            "total_frames": total_frames,
            "duration": duration
        }
        
    except Exception as e:
        return {"error": str(e)}


def main():
    parser = argparse.ArgumentParser(description='Detect scenes in a video')
    parser.add_argument('video_path', help='Path to the video file')
    parser.add_argument('--threshold', type=float, default=20.0, help='Detection threshold (0-255), lower is more sensitive')
    parser.add_argument('--min-scene-length', type=float, default=1.0, help='Minimum scene length in seconds')
    parser.add_argument('--fps', type=float, default=None, help='Video FPS (optional, for reference)')
    
    args = parser.parse_args()
    
    result = detect_scenes(args.video_path, args.threshold, args.min_scene_length)
    
    # Output as JSON
    print(json.dumps(result))


if __name__ == '__main__':
    main()

File: services/sceneDetector.js
const { spawn } = require('child_process');
const path = require('path');
const ffmpeg = require('./ffmpeg');

const DEFAULT_MIN_SCENE_LENGTH = 1.0; // seconds
const DEFAULT_THRESHOLD = 20.0; // Lower is more sensitive (0-255)
const PYTHON_SCRIPT = path.join(__dirname, 'scene_detector.py');

/**
 * Detect scenes using Python pyscenedetect
 * @param {string} inputPath - Path to video file
 * @param {Object} options - Detection options
 * @returns {Promise<Array>} Array of scene objects
 */
async function detectScenes(inputPath, options = {}) {
  const videoInfo = await ffmpeg.getVideoInfo(inputPath);
  const fps = options.fps || videoInfo.fps;
  
  // Min scene length in seconds (from slider)
  let minSceneLengthSec = options.minSceneLength || DEFAULT_MIN_SCENE_LENGTH;
  
  // If value seems to be in frames (> 300), convert to seconds
  if (minSceneLengthSec > 300) {
    minSceneLengthSec = minSceneLengthSec / fps;
  }
  
  // Clamp to reasonable range (0.5 - 60 seconds)
  minSceneLengthSec = Math.max(0.5, Math.min(60, minSceneLengthSec));
  
  // Threshold for pyscenedetect (0-255, lower is more sensitive)
  const threshold = options.threshold || DEFAULT_THRESHOLD;
  
  console.log(`Detecting scenes with pyscenedetect: threshold=${threshold}, minLength=${minSceneLengthSec}s, fps=${fps}`);
  
  return new Promise((resolve, reject) => {
    const pythonProcess = spawn('python3', [
      PYTHON_SCRIPT,
      inputPath,
      '--threshold', threshold.toString(),
      '--min-scene-length', minSceneLengthSec.toString(),
      '--fps', fps.toString()
    ]);
    
    let output = '';
    let errorOutput = '';
    
    pythonProcess.stdout.on('data', (data) => {
      output += data.toString();
    });
    
    pythonProcess.stderr.on('data', (data) => {
      errorOutput += data.toString();
    });
    
    pythonProcess.on('close', (code) => {
      if (code !== 0) {
        console.error('Python scene detection error:', errorOutput);
        // Fallback to FFmpeg scene detection
        console.log('Falling back to FFmpeg scene detection...');
        fallbackDetectScenes(inputPath, minSceneLengthSec, fps)
          .then(resolve)
          .catch(reject);
        return;
      }
      
      try {
        const result = JSON.parse(output);
        
        if (result.error) {
          console.error('Scene detection error:', result.error);
          // Fallback to FFmpeg
          fallbackDetectScenes(inputPath, minSceneLengthSec, fps)
            .then(resolve)
            .catch(reject);
          return;
        }
        
        resolve(result.scenes || []);
      } catch (err) {
        console.error('Failed to parse Python output:', err);
        // Fallback to FFmpeg
        fallbackDetectScenes(inputPath, minSceneLengthSec, fps)
          .then(resolve)
          .catch(reject);
      }
    });
    
    pythonProcess.on('error', (err) => {
      console.error('Failed to start Python process:', err);
      // Fallback to FFmpeg
      fallbackDetectScenes(inputPath, minSceneLengthSec, fps)
        .then(resolve)
        .catch(reject);
    });
  });
}

/**
 * Fallback scene detection using FFmpeg
 */
async function fallbackDetectScenes(inputPath, minSceneLengthSec, fps) {
  console.log('Using FFmpeg fallback for scene detection');
  const minSceneLengthFrames = Math.round(minSceneLengthSec * fps);
  const scenes = await ffmpeg.detectScenesFFmpeg(inputPath, 0.3, minSceneLengthFrames, fps);
  
  // Get video info for validation
  const videoInfo = await ffmpeg.getVideoInfo(inputPath);
  
  // Validate and clean up scene data
  return scenes.map((scene, index) => ({
    scene_number: index + 1,
    start_frame: Math.max(1, Math.round(scene.start_frame)),
    end_frame: Math.min(videoInfo.total_frames, Math.round(scene.end_frame)),
    start_time: Math.max(0, parseFloat(scene.start_time.toFixed(3))),
    end_time: Math.min(videoInfo.duration, parseFloat(scene.end_time.toFixed(3))),
    duration: parseFloat((scene.end_time - scene.start_time).toFixed(3))
  }));
}

/**
 * Validate frame range for extraction
 * @param {number} startFrame - Start frame number
 * @param {number} endFrame - End frame number
 * @param {number} totalFrames - Total frames in video
 * @returns {Object} Validation result
 */
function validateFrameRange(startFrame, endFrame, totalFrames) {
  const errors = [];
  
  if (!Number.isInteger(startFrame) || startFrame < 1) {
    errors.push('Start frame must be a positive integer');
  }
  
  if (!Number.isInteger(endFrame) || endFrame < 1) {
    errors.push('End frame must be a positive integer');
  }
  
  if (startFrame > endFrame) {
    errors.push('Start frame must be less than or equal to end frame');
  }
  
  if (endFrame > totalFrames) {
    errors.push(`End frame cannot exceed total frames (${totalFrames})`);
  }
  
  const frameCount = endFrame - startFrame + 1;
  if (frameCount > 1000) {
    errors.push('Cannot extract more than 1000 frames at once');
  }
  
  return {
    valid: errors.length === 0,
    errors,
    frameCount
  };
}

/**
 * Calculate frame range with offset and length
 * @param {number} startFrame - Original start frame
 * @param {number} endFrame - Original end frame
 * @param {number} offset - Frame offset (can be negative)
 * @param {number} length - Desired length in frames
 * @param {number} totalFrames - Total frames in video
 * @returns {Object} New frame range
 */
function calculateFrameRange(startFrame, endFrame, offset, length, totalFrames) {
  // Apply offset
  let newStart = startFrame + offset;
  let newEnd = endFrame + offset;
  
  // Apply length constraint if specified
  if (length && length > 0) {
    newEnd = newStart + length - 1;
  }
  
  // Clamp to valid range
  newStart = Math.max(1, newStart);
  newEnd = Math.min(totalFrames, newEnd);
  
  // Ensure start <= end
  if (newStart > newEnd) {
    newEnd = newStart;
  }
  
  return {
    start_frame: newStart,
    end_frame: newEnd,
    frame_count: newEnd - newStart + 1
  };
}

module.exports = {
  detectScenes,
  validateFrameRange,
  calculateFrameRange,
  DEFAULT_MIN_SCENE_LENGTH
};